### DISCUSSION TITLE
audio_process_extern seems to give me 1/2 the data....

### DISCUSSION THREAD

**Discussion by themindfactory** [schreibfaul1/ESP32-audioI2S#722] [Q&A] [answered]
I am intercepting the decoded mp3 data to do some FFT work with it... however I have noticed it seems to be sending 1/2 the data... 

I tried this to test:
void audio_process_extern(int16_t* buff, uint16_t len, bool *continueI2S) {
  Serial.printf("a %d %d\n", len, millis());
  *continueI2S = true;
}

its sending me len == 1152 every 25ms approx, so that is almost exactly the 44100 sample rate of the mp3 being played, however I would expect 2X that data as its stereo, and I have set forcemono to false, what am I not understanding? did I miss something?

The MP3 is also from a stream online....

Richard.

**Comment by schreibfaul1** [comment]
Exactly, if the source is a stereo transmitter, the right/left channels are transmitted alternately.
"forcemono" actually only generates a mixed signal which is then transmitted right/left. This is required if there is only one loudspeaker.
The situation is different if the source is mono, in which case half of the samples are available.

**Reply by themindfactory** [reply]
I had said forcemono is false, source is stereo and i2s device is stereo, so I expect 2x more data

**Reply by schreibfaul1** [reply]
forcemono has no influence, audio_process_extern() provides the raw data of the decoder.

**Reply by themindfactory** [reply]
So then I am still confused why when streaming a stereo mp3 file it gives me 1/2 the data... 

info        Channels: 2
info        SampleRate: 44100
info        BitsPerSample: 16
info        BitRate: 128000

```
My test code:
void audio_process_extern(int16_t* buff, uint16_t len, bool *continueI2S) {
  static int sampleCnt = 0;

  sampleCnt += len;
  EVERYMS(1000) {
    Serial.printf("len %d\n", sampleCnt);
    sampleCnt = 0;
  }
}

```

which prints out 44100 every second.... would this not be 88200 for stereo?

Thanks!

**Reply by themindfactory** [reply]
or is the size of buff 2 x len (int16_t), then I understand, I expected len to be the len of the int16_t buffer

**Reply by schreibfaul1** [reply, marked as answer]
Yes:   numSamples = len * audio.getChannels()

**Reply by themindfactory** [reply]
thanks for the library its working very nicely... :-)
